AEP: `CalcJob` live monitoring

sphuber opened the pull request:
Fixes #1925

Implementation of [AEP: Allow CalcJobs to be actively monitored and interrupted](https://github.com/aiidateam/AEP/pull/36)

sphuber commented:
@giovannipizzi @ramirezfranciscof let me know if you would like to review this still. I would like to get this merged soon for the v2.1 releaes

ramirezfranciscof commented:
Sorry @sphuber , busy weeks. Now I need to prepare for my GM on friday, would it be ok if I take a look on friday afternoon?

ramirezfranciscof commented:
Hey, I'm still working on testing this, but one issue I found is that this seems to be using a type operation that is only supported from python 3.10 on ([source](https://github.com/tiangolo/typer/issues/371)). When running on 3.9 I get the following (in 3.10 it works fine):

```
Traceback (most recent call last):
  File "/home/framirez/miniconda3/envs/aurora-aiida/bin/verdi", line 8, in <module>
    sys.exit(verdi())
  File "/home/framirez/miniconda3/envs/aurora-aiida/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/framirez/miniconda3/envs/aurora-aiida/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/framirez/miniconda3/envs/aurora-aiida/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/framirez/miniconda3/envs/aurora-aiida/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/framirez/miniconda3/envs/aurora-aiida/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/framirez/Workenvs/aurora/aiida-core/aiida/cmdline/utils/decorators.py", line 73, in wrapper
    return wrapped(*args, **kwargs)
  File "/home/framirez/Workenvs/aurora/aiida-core/aiida/cmdline/commands/cmd_run.py", line 115, in run
    exec(compile(handle.read(), str(filepath), 'exec', dont_inherit=True), globals_dict)  # pylint: disable=exec-used
  File "submit_newversion.py", line 68, in <module>
    cycler_builder.monitors = {
  File "/home/framirez/Workenvs/aurora/aiida-core/aiida/engine/processes/builder.py", line 123, in __setattr__
    raise ValueError(f'invalid attribute value {validation_error.message}')
ValueError: invalid attribute value `monitors.testing_monitor` is invalid: unsupported operand type(s) for |: 'type' and 'type'
```

sphuber commented:
Thanks for the report @ramirezfranciscof . It's interesting that it is failing because the CI runs on Python 3.8 and the tests pass just fine. Can you share the script that you are running? Note that if you use this notation in older Python versions, you need to explicitly enable it with `from __future__ import annotations`. Maybe you are using this type notation (`str | None`, for example) in your test script and that is why it is complaining.

ramirezfranciscof commented:
Yeap, you are right. It is not exactly on the test script but on the plugin implementation, as I was copying the examples in the [AEP](https://github.com/aiidateam/AEP/blob/22d17814f73171b32a7577297ed788fbca74b803/008_calcjob_monitors/readme.md#base-functionality) and didn't realize they were using this python 10 feature.

ramirezfranciscof reviewed (changes requested):
Hey @sphuber apologies for the delay, but I can confirm that the feature seems to be compatible with what we are trying to do with the Aurora project. Here are my comments on the code.

One general issue: whenever a calcjob finishes correctly, the status gets stuck in `Waiting for transport task: retrieve`. Also the kill command leaves a `None` status rather than the `Killed through 'verdi process kill'`:

```
  PK  Created    Process label            Process State     Process status
----  ---------  -----------------------  ----------------  ------------------------------------
 161  5D ago     datanode_preparation     ⏹ Finished [0]
 165  5D ago     BatteryCyclerExperiment  ☠ Killed          Killed through `verdi process kill`
 168  5D ago     CalcjobMonitor           ⏹ Finished [0]
 176  5D ago     datanode_preparation     ⏹ Finished [0]
 180  5D ago     BatteryCyclerExperiment  ☠ Killed          None
 183  5D ago     CalcjobMonitor           ⏹ Finished [0]    Waiting for transport task: retrieve
 186  5D ago     datanode_preparation     ⏹ Finished [0]
 190  5D ago     BatteryCyclerExperiment  ⏹ Finished [0]    Waiting for transport task: retrieve
 218  1D ago     datanode_preparation     ⏹ Finished [0]
 223  1D ago     BatteryCyclerExperiment  ⏹ Finished [0]    Waiting for transport task: retrieve
 274  23h ago    datanode_preparation     ⏹ Finished [0]
 279  23h ago    BatteryCyclerExperiment  ⏹ Finished [150]
 284  22h ago    datanode_preparation     ⏹ Finished [0]
 289  22h ago    BatteryCyclerExperiment  ⏹ Finished [150]
 334  21h ago    datanode_preparation     ⏹ Finished [0]
 339  21h ago    BatteryCyclerExperiment  ⏹ Finished [0]    Waiting for transport task: retrieve
 344  5h ago     datanode_preparation     ⏹ Finished [0]
 349  5h ago     BatteryCyclerExperiment  ⏹ Finished [150]
 354  4h ago     datanode_preparation     ⏹ Finished [0]
 358  4h ago     BatteryCyclerExperiment  ⏹ Finished [0]    Waiting for transport task: retrieve
 363  4h ago     datanode_preparation     ⏹ Finished [0]
 367  4h ago     BatteryCyclerExperiment  ☠ Killed          None
 370  4h ago     CalcjobMonitor           ⏹ Finished [0]    Waiting for transport task: retrieve

Total results: 23
```

(Calculation 165 was run with a previous version of aiida)

I think this might be related to moving the `set_process_status` method [out of the try statement](https://github.com/aiidateam/aiida-core/pull/5659/files#diff-e599775aef84fefc9a259d852070235f0328c8db135b335d85f176aeaf02c0acR486) but this is just a guess.

ramirezfranciscof commented on aiida/engine/processes/calcjobs/monitors.py:
```diff
@@ -0,0 +1,196 @@
+    action: CalcJobMonitorAction = CalcJobMonitorAction.KILL
+    """The action the engine should take."""
+
+    retrieve: bool = True
+    """If set to ``False``, the engine will skip retrieving the output files."""
+
+    parse: bool = True
+    """If set to ``False``, the engine will skip the parsing of the retrieved files, if one was specified in inputs."""
```
Mmm but I guess you wanted to leave it open to enable retrieve/parse depending on the reason why the calcjob is being stopped, which I guess makes sense. Nvm then, I answered myself on this one.

sphuber commented on aiida/engine/processes/calcjobs/monitors.py:
```diff
@@ -0,0 +1,196 @@
+        :raises TypeError: If any of the attributes are of the incorrect type.
+        :raises ValueError: If ``parse == True`` and ``retrieve == False``.
+        """
+        type_check(self.key, str, allow_none=True)
+        type_check(self.message, str, allow_none=True)
+        type_check(self.action, CalcJobMonitorAction)
+        type_check(self.retrieve, bool)
+        type_check(self.parse, bool)
+        type_check(self.override_exit_code, bool)
```
I did, but I don't want to add a dependency just for this.

ramirezfranciscof replied:
Fair enough. I think it might be worth considering to do more type checking in general in other sections of the code or move the ones that already exist to Pydantic, so it wouldn't be just for this, but I guess that can be a separate issue.

sphuber commented on aiida/engine/processes/calcjobs/calcjob.py:
```diff
@@ -516,8 +535,20 @@ def parse(self, retrieved_temporary_folder: Optional[str] = None) -> ExitCode:
+
         return exit_code or ExitCode(0)
 
+    @staticmethod
+    def terminate(exit_code: ExitCode) -> ExitCode:
+        """Terminate the process immediately and return the given exit code.
+
+        :param exit_code: The exit code to return.
+        :returns: The provided exit code.
+        """
+        return exit_code
```
This is called by the `Waiting.execute` method if a monitor triggers and sets `parse=false`. Normally, the execute method recreates the `RUNNING` state with the `parse` method to be called, but here we need another method that will simply terminate the process. Any process step that returns an exit code will cause it to be terminated, so that is what this hook allows us to do. There was no other process step that was a no-op like this (I don't believe).

sphuber replied:
I have updated the docstring to explain this.

sphuber commented:
Thanks for testing and reviewing the code @ramirezfranciscof 

> One general issue: whenever a calcjob finishes correctly, the status gets stuck in `Waiting for transport task: retrieve`.

That's a bug indeed. Will have a look to fix it.

> Also the kill command leaves a `None` status rather than the `Killed through 'verdi process kill'`:

I don't think it should show `Killed through 'verdi process kill'` though because, well, that's not true :smile: We could show something else, but since we already have the dedicated exit code `150` I thought this wasn't necessary.

ramirezfranciscof commented:
> I don't think it should show `Killed through 'verdi process kill'` though because, well, that's not true 😄 We could show something else, but since we already have the dedicated exit code `150` I thought this wasn't necessary.

Haha, yes that is true. I guess what I meant is that the `None` status is a bit atypical, shouldn't it just show nothing like in regular finished calcjobs? Although now that I think of it, is there any drawback to showing something like `Killed by monitor <monitor_keyname>` in the status? Or just `Killed by monitor` if these status can't be "personalized".

sphuber commented on aiida/engine/processes/calcjobs/tasks.py:
```diff
@@ -220,6 +222,60 @@ async def do_update():
+            transport = await cancellable.with_interrupt(request)
+            transport.chdir(node.get_remote_workdir())
+
+            monitors = CalcJobMonitors(node.inputs.monitors)
+            return monitors.process(node, transport, disabled_monitors)
```
I think you are right and this is a bug that is not covered by the unit test, good catch. I will see if I can write the unit test in a way that it doesn't maintain its own monitors collection or add an integration test.

sphuber commented:
> Although now that I think of it, is there any drawback to showing something like Killed by monitor <monitor_keyname> in the status? Or just Killed by monitor if these status can't be "personalized".

We could. We could set the process status to the message of the exit code, which does contain the specific monitor error, but the thing is that it is not really consistent as we also don't do this for any other exit code. In my mind, the code being killed by a monitor is just like the parser returning a non-zero exit code. If a user sees `[150]` they can run `verdi process show` or `node.exit_message` in the API to get the info on the particular monitor that triggered.

sphuber commented:
> Haha, yes that is true. I guess what I meant is that the `None` status is a bit atypical, shouldn't it just show nothing like in regular finished calcjobs?

Ah my bad, I now see what you mean, it is when you actually kill a job it shows literal `None` in the process status. That would be a regression bug yeah. Will look into it.

sphuber commented:
@ramirezfranciscof thanks again for the thorough testing. I have added two commits for the bugs that you uncovered. I added them on top so it is easier for you to review the changes. Once you sign off on the changes, I will rebase them into the proper commits.
